### PR TITLE
 Improved P256 ECC performance by means of a dedicated function for modular inversion modulo the P256 group order

### PR DESCRIPTION
**Pull request by vkrasnov**
PR for: https://rt.openssl.org/Ticket/Display.html?id=3810 

The patch improves upon our previous submission, by providing a dedicated function to perform modular inversion modulo the P256 group order.

Results:
The performance improvements, for single threaded applications, compared to the current (development) version of OpenSSL are as follows.

(measured by "openssl speed" utility)
On Haswell:
ECDSA sign: 1.28X
ECDSA verify: 1.10X

On Broadwell:
ECDSA sign: 1.42X
ECDSA verify: 1.18X

### CODE REVIEW

**Comment by Adamzynoni** [comment]
how can i create account?

**Comment by briansmith** [comment]
I noticed that this code is almost 100% self contained (except for the use of  ecp_nistz256_bignum_to_field_elem, if you follow my last suggestion) and separated from the rest of the existing ecp_nistz256 code. I also noticed that it can (should) be used in conjunction with the other implementations like ecp_nistp256.c. Wouldn't it be better to separate it (both the ASM code and the C code) to make that easier to do? In particular, some applications may decide to disable the ecp_nistz256 code that uses the 140KB table for space reasons, but may want this optimized inversion implementation.

Also, is there a paper to go along with this submission. I would like to read it, if so.

**Comment by vkrasnov** [comment]
> Also, is there a paper to go along with this submission. I would like to read it, if so.

No paper on that, but it mostly reuses the same optimizations described in the original paper, but for the order, which also has special form.
So you can do a faster reduction, and save many multiplications in the modular exponentiation.

**Comment by mattcaswell** [comment]
Ping @dot-asm, @ekasper. Any opinions on this PR?

**Comment by ekasper** [comment]
The licensing issues for this PR have not been cleared still.

It appears to be way too late for 1.1.0 now...

**Comment by vkrasnov** [comment]
Wow, it has been a year, almost forgot about that one :)
I can make it available under any license really, although it is just improvement over the existing code, so might not need any license.

**Comment by mattcaswell** [comment]
@ekasper yes, agreed - too late for 1.1.0. But if @vkrasnov made it available under the OpenSSL licence would it be acceptable?

**Comment by richsalz** [comment]
Better to have NO license so we can move to ASF. And a CLA

**Comment by vkrasnov** [comment]
So should I go ahead and rebase it etc.?

**Comment by ekasper** [comment]
Sorry, I did not realise this was from _last_ April! :(

@vkrasnov, thanks for not giving up on us - yes, this looks like a great improvement. Please do rebase; I will try to leave some quick comments now, and then let's pick it up right after 1.1.0.

One particular request: I'd very much appreciate math / pseudocode comments to guide the reader through the asm. Coincidentally I just did this for curve25519 inversion so you can look at 0eadff033fea00f1b9abe0a83bf0d6637690f085 for inspiration :)

File: include/openssl/ec.h [hunk 1]
```diff
@@ -248,6 +248,17 @@ const EC_POINT *EC_GROUP_get0_generator(const EC_GROUP *group);
...
+/** Perfroms inversion modulo order of a EC_GROUP
+ *  \param group  EC_GROUP object
+ *  \param res    BIGNUM to which the result is copied
+ *  \param x      BIGNUM of which inverse to compute
+ *  \param ctx    BN_CTX object
+ *  \param constantTime int   if 0 - not constant time
+ *  \return 1 on success and 0 if an error occurred
+*/
+int EC_GROUP_do_inverse_ord(const EC_GROUP *group, BIGNUM *res,
+                            const BIGNUM *x, BN_CTX *ctx, int constantTime);
```
Comment (ekasper) [inline include/openssl/ec.h]: Is constantTime = 0 worth it? I'd much prefer not to even have this option, esp exposed in the public API.

File: include/openssl/ec.h [hunk 2]
```diff
@@ -248,6 +248,17 @@ const EC_POINT *EC_GROUP_get0_generator(const EC_GROUP *group);
...
 
+/** Perfroms inversion modulo order of a EC_GROUP
+ *  \param group  EC_GROUP object
+ *  \param res    BIGNUM to which the result is copied
+ *  \param x      BIGNUM of which inverse to compute
+ *  \param ctx    BN_CTX object
+ *  \param constantTime int   if 0 - not constant time
+ *  \return 1 on success and 0 if an error occurred
+*/
+int EC_GROUP_do_inverse_ord(const EC_GROUP *group, BIGNUM *res,
```
Comment (ekasper) [inline include/openssl/ec.h]: Perhaps just `EC_GROUP_inverse_mod_order` or `EC_GROUP_invert_mod_order`?

File: crypto/ec/ecp_nistz256.c [hunk 3]
```diff
@@ -1519,7 +1664,12 @@ const EC_METHOD *EC_GFp_nistz256_method(void)
         0,                                          /* field_div */
         ec_GFp_mont_field_encode,
         ec_GFp_mont_field_decode,
-        ec_GFp_mont_field_set_to_one
+        ec_GFp_mont_field_set_to_one,
+#if defined(__x86_64) || defined(__x86_64__)
```
Comment (ekasper) [inline crypto/ec/ecp_nistz256.c]: This glue does not look quite right - I can imagine at some point we may have this for other than x86_64, too. Perhaps we can integrate a define (`ECP_NISTZ256_ASM_MOD_INVERSE` or some such) into the Configure cascades? (@dot-asm may have a suggestion?)

Also, drive-by and I haven't really thought about it myself but wdyt about the suggestion to disentangle this from nistz and engage it for ecp_nistp256 as well? Worth it?

File: crypto/ec/ecp_nistz256.c [hunk 4]
```diff
@@ -1357,9 +1402,109 @@ __owur static int ecp_nistz256_points_mul(const EC_GROUP *group,
...
+    ret = 1;
+err:
+    return ret;
+}
+#endif
+
 __owur static int ecp_nistz256_get_affine(const EC_GROUP *group,
-                                          const EC_POINT *point,
-                                          BIGNUM *x, BIGNUM *y, BN_CTX *ctx)
+                                   const EC_POINT *point,
```
Comment (ekasper) [inline crypto/ec/ecp_nistz256.c]: Nit: the C has wrong indentation/spacing/style in places - if you could do a quick pass while rebasing, that'd be great. E.g. `lowercase_with_underscore` naming, spaces around assignment and comparison `a *= 2`, no spaces inside parens `(i = 0; i < 32; i++)` etc. (coding style: https://www.openssl.org/policies/codingstyle.html).

File: include/openssl/ec.h [hunk 5]
```diff
@@ -248,6 +248,17 @@ const EC_POINT *EC_GROUP_get0_generator(const EC_GROUP *group);
 */
 BN_MONT_CTX *EC_GROUP_get_mont_data(const EC_GROUP *group);
 
+/** Perfroms inversion modulo order of a EC_GROUP
```
Comment (ekasper) [inline include/openssl/ec.h]: Typo: performs. Also update the manpages, please.

File: crypto/ec/ecp_nistz256.c [hunk 6]
```diff
@@ -1357,9 +1402,109 @@ __owur static int ecp_nistz256_points_mul(const EC_GROUP *group,
...
+                                    BIGNUM *r,
+                                    const BIGNUM *x,
+                                    BN_CTX *ctx)
+{
+    /* RR = 2^512 mod ord(p256) */
+    static const BN_ULONG RR[P256_LIMBS] = {TOBN(0x83244c95,0xbe79eea2),
+                                            TOBN(0x4699799c,0x49bd6fa6),
+                                            TOBN(0x2845b239,0x2b6bec59),
+                                            TOBN(0x66e12d94,0xf3d95620)};
+    /* The constant 1 (unlike ONE that is one in Montgomery representation) */
```
Comment (ekasper) [inline crypto/ec/ecp_nistz256.c]: Perhaps ONE should be renamed MONT_ONE instead?

File: crypto/ec/ecp_nistz256.c [hunk 7]
```diff
@@ -1357,9 +1402,109 @@ __owur static int ecp_nistz256_points_mul(const EC_GROUP *group,
...
+    ecp_nistz256_bignum_to_field_elem(out, x);
+    ecp_nistz256_ord_mul_mont(&table[0*P256_LIMBS], out, RR);
+    for ( i = 2; i < 16; i+=2 ) {
+        ecp_nistz256_ord_sqr_mont(&table[(i-1)*P256_LIMBS],
+                                  &table[(i/2-1)*P256_LIMBS], 1);
+        ecp_nistz256_ord_mul_mont(&table[i*P256_LIMBS],
+                                  &table[(i-1)*P256_LIMBS],
+                                  &table[0*P256_LIMBS]);
+    }
+    /* The top 128bit of the exponent are highly redundant,
```
Comment (ekasper) [inline crypto/ec/ecp_nistz256.c]: What does "highly redundant" mean here? Do you mean "highly regular", i.e, that you optimise for the `ffff...0000....` pattern?

File: crypto/ec/ecp_nistz256.c [hunk 8]
```diff
@@ -1357,9 +1402,109 @@ __owur static int ecp_nistz256_points_mul(const EC_GROUP *group,
...
+    int i, ret = 0;
+    BIGNUM *tmp;
+
+    if ((BN_num_bits(x) > 256)
+        || BN_is_negative(x)) {
+        if ((tmp = BN_CTX_get(ctx)) == NULL) {
+            ECerr(EC_F_ECP_NISTZ256_INV_ORD, ERR_R_BN_LIB);
+            goto err;
+        }
+        if (!BN_nnmod(tmp, x, group->order, ctx)) {
```
Comment (ekasper) [inline crypto/ec/ecp_nistz256.c]: Needs a comment about violating constant-timeness? (This is fine, IIRC we do the same in the ecp_nist modules for out-of-range inputs - but worth a comment.)

**Comment by dot-asm** [comment]
And it was pending in internal review system for almost 6 months... With a number of small issues addressed. Like above mentioned `#if` condition, elimination of redundant memcpy, minor optimizations to assembly code, as well as annotations...

**Comment by richsalz** [comment]
@dot-asm I just plus-one'd your version.

**Comment by romen** [comment]
It seems like this could be closed:

- `ecp_nistz256` does have a custom inversion (see e.g., <https://github.com/openssl/openssl/blob/master/crypto/ec/ecp_nistz256.c#L1300>)
- other PRs got merged that likely included the changes in this PR (see e.g., #5001)

Thanks @bbbrumley for a second pair of eyes in checking the current state of this PR and quickly digging related PRs!

**Comment by paulidale** [comment]
+1 for closing

**Pull request closed without merging**